File: pydantic_agent/src/agent_factory.py
from __future__ import annotations
from pydantic_ai import Agent
from pydantic_ai.models.openai import OpenAIModel
from pydantic_ai.providers.openai import OpenAIProvider
from .bff_tool_adapter import build_tool_functions
from .models import BffDeps


def build_agent(
    tool_schemas: list[dict],
    model_name: str,
    base_url: str,
    api_key: str,
    system_prompt: str | None = None,
) -> Agent:
    # Constructing OpenAIModel with an explicit provider keeps pydantic_ai from
    # falling back to the env-driven default (which would 401 on LM Studio).
    model = OpenAIModel(
        model_name=model_name,
        provider=OpenAIProvider(base_url=base_url, api_key=api_key),
    )
    tools = build_tool_functions(tool_schemas)
    prompt = system_prompt or (
        "You are a helpful banking assistant. "
        "Use the available tools to answer user questions accurately."
    )
    return Agent(
        model,
        deps_type=BffDeps,
        tools=tools,
        system_prompt=prompt,
        defer_model_check=True,
    )

File: mastra_agent/src/agentFactory.ts
import { Agent } from '@mastra/core/agent';
import { createOpenAI } from '@ai-sdk/openai';
import { buildBffTools, type ToolSchema, type RunCtx } from './bffToolAdapter';

const DEFAULT_INSTRUCTIONS =
  'You are a helpful banking assistant. Use the available tools to answer the user\'s questions.';

export interface LlmProviderConfig {
  baseUrl: string;
  apiKey: string;
  model: string;
}

export function buildAgent(
  toolSchemas: ToolSchema[],
  runCtx: RunCtx,
  llm: LlmProviderConfig,
  instructions?: string,
): Agent {
  const tools = buildBffTools(toolSchemas, runCtx);

  const toolMap: Record<string, (typeof tools)[number]> = {};
  for (const tool of tools) {
    toolMap[tool.id] = tool;
  }

  // createOpenAI() with a baseURL points the OpenAI provider at any
  // OpenAI-compatible endpoint (LM Studio, Groq, Together, etc).
  const provider = createOpenAI({ baseURL: llm.baseUrl, apiKey: llm.apiKey });

  // provider(modelId) returns an AI SDK LanguageModel that Mastra's Agent
  // accepts at runtime, but its TS surface is DynamicArgument<MastraModelConfig>
  // which doesn't include the AI SDK type directly — cast through unknown.
  return new Agent({
    id: 'banking-agent',
    name: 'Banking Agent',
    instructions: instructions ?? DEFAULT_INSTRUCTIONS,
    model: provider(llm.model) as unknown as never,
    tools: toolMap,
  });
}
